libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,581 @@
+  Vector *v,
+  char **pzErrMsg
+){
+  switch( sqlite3_value_type(arg) ){
+    case SQLITE_NULL:
+      *pzErrMsg = sqlite3_mprintf("invalid vector: NULL");
+      return -1;
+    case SQLITE_BLOB:
+      return vectorParseBlob(arg, v, pzErrMsg);
+    case SQLITE_TEXT:
+      return vectorParseText(arg, v, pzErrMsg);
+    default:
+      *pzErrMsg = sqlite3_mprintf("invalid vector: not a text or blob type");
+      return -1;
+  }
+}
+
+static inline int isInteger(float num){
+  return num == (u64)num;
+}
+
+void vectorDump(Vector *pVec){
+  switch (pVec->type) {
+    case VECTOR_TYPE_FLOAT32:
+      vectorF32Dump(pVec);
+      break;
+    case VECTOR_TYPE_FLOAT64:
+      vectorF64Dump(pVec);
+      break;
+    default:
+      assert(0);
+  }
+}
+
+static void vectorDeserialize(
+  sqlite3_context *context,
+  Vector *v
+){
+  switch (v->type) {
+    case VECTOR_TYPE_FLOAT32:
+      vectorF32Deserialize(context, v);
+      break;
+    case VECTOR_TYPE_FLOAT64:
+      vectorF64Deserialize(context, v);
+      break;
+    default:
+      assert(0);
+  }
+}
+
+static void vectorSerialize(
+  sqlite3_context *context,
+  Vector *v
+){
+  if( v->dims==0 ) {
+    sqlite3_result_null(context);
+    return;
+  }
+  switch (v->type) {
+    case VECTOR_TYPE_FLOAT32:
+      vectorF32Serialize(context, v);
+      break;
+    case VECTOR_TYPE_FLOAT64:
+      vectorF64Serialize(context, v);
+      break;
+    default:
+      assert(0);
+  }
+}
+
+size_t vectorSerializeToBlob(Vector *p, unsigned char *blob, size_t blobSize){
+  switch (p->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32SerializeToBlob(p, blob, blobSize);
+    case VECTOR_TYPE_FLOAT64:
+      return vectorF64SerializeToBlob(p, blob, blobSize);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+size_t vectorDeserializeFromBlob(Vector *p, const unsigned char *blob, size_t blobSize){
+  switch (p->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32DeserializeFromBlob(p, blob, blobSize);
+    case VECTOR_TYPE_FLOAT64:
+      
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+/**************************************************************************
+** Vector index cursor implementations
+****************************************************************************/
+
+/*
+** A VectorIdxCursor is a special cursor to perform vector index lookups.
+ */
+struct VectorIdxCursor {
+  sqlite3 *db;          /* Database connection */
+  DiskAnnIndex *index;   /* DiskANN index on disk */
+};
+
+/**
+** Parses a type string such as `FLOAT32(3)` and returns the number of dimensions.
+**
+** Returns -1 if the type string is not a valid vector type.
+**/
+static int parseVectorDims(const char *zType){
+  int dims = 0;
+  if( sqlite3_strnicmp(zType, "FLOAT32(", 8)==0 || sqlite3_strnicmp(zType, "FLOAT64(", 8)==0){
+    const char *z = zType + 8;
+    while( *z && *z!=')' ){
+      if( !sqlite3Isdigit(*z) ){
+        return -1;
+      }
+      dims = dims*10 + (*z - '0');
```

Comment:
Is it possible to define FLOAT32(X) with X being bigger than MAX_INT? Would it make sense to check for overflow here and report user readable error explaining that they try to use too big number of dimensions in case of X > MAX_INT or even X > MAX_VECTOR_DIMS?